src/ListT/Weighted.hs

```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | A re-implementation of @weighted-search@
+-- as a monad transformer.
+module ListT.Weighted
+  ( Weight (..),
+    WeightedT,
+    Weighted,
+    weight,
+    (<|>),
+    mapWeightedT,
+    toList,
+    firstThat,
+  )
+where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus (..), ap)
+import Control.Monad.Error.Class (MonadError (..))
+import Control.Monad.IO.Class (MonadIO (..))
+import Control.Monad.Reader.Class (MonadReader (..))
+import Control.Monad.State.Class (MonadState (..))
+import Control.Monad.Trans (MonadTrans (..))
+import Control.Monad.WeightedSearch (Weight (..))
+import Data.Foldable (fold)
+import Data.Functor.Identity (Identity)
+
+-- | Weighted nondeterminstic computations over the weight @w@.
+data WeightedT w m a
+  = Fail
+  | -- | returns one value, and then more of them
+    Yield a (WeightedT w m a)
+  | -- | signals that the internal computation is weighted
+    Weight w (WeightedT w m a)
+  | -- | perform an action
+    Action (m (WeightedT w m a))
```

Comment:
It seems we could also define this as:
```suggestion
data WeightedT w m a
  = Empty
  | -- | returns one value, and then more of them
    Yield (m a) (m (WeightedT w m a))
  | -- | signals that the internal computation is weighted
    Weight w (WeightedT w m a)
```
I wonder about the tradeoffs :thinking:

Reply:
I started with that, but I didn't manage to implement `mplus`. So I turned into this other representation.

Reply:
Fair! All of this work because we want to burry an IO monad somewhere. Makes me wonder whether we should just `unsafePerformIO` the heck out of the SMT interface

Reply:
This typechecks:

```haskell
instance (Alternative m) => Alternative (T m) where
  Fail <|> r = r
  r <|> Fail = r
  Yield ma mr <|> r = Yield ma (flip (<|>) r <$> mr)
  Weight w r <|> Yield mb ms = Yield mb ((Weight w r <|>) <$> ms)
  Weight w r <|> Weight w' s =
    case compare w w' of
      EQ -> Weight w  (r <|> s)
      LT -> Weight w  (r <|> Weight (w' - w) s)
      GT -> Weight w' (r <|> Weight (w - w') s)
  ```

The advantage I see here is that it also orders the execution of effects; our current representation with `Action` only orders the results, but still executes the actions in the order it encounters them, no? I could be totally wrong though, so don't quote me on that!

Reply:
Now I cannot get to implement `>>=` :thinking:  The problem is that we have:

```haskell
Yield x m >>= f = mplus (f x) <$> ((>>= f) <$> m)
```

But how do I know whether I have a `Yield` or something else to put on the top of that?

Reply:
hmm... Maybe we play the `ListT` trick:

```haskell
data T m a
  = Yield (m (a, T m a))
  | Weight Natural (T m a)
```

I'd assume that now we can copy the bind from `ListT`

Reply:
But now I think this is the same as you've got...